server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/fileformats/data/StarTreeDataWriter.java

```diff
@@ -66,21 +66,32 @@ private static void writeStarTreeNodes(IndexOutput output, InMemoryTreeNode root
         while (!queue.isEmpty()) {
             InMemoryTreeNode node = queue.remove();
 
-            if (node.children == null || node.children.isEmpty()) {
+            if ((node.children == null || node.children.isEmpty()) && node.childStarNode == null) {
                 writeStarTreeNode(output, node, ALL, ALL);
             } else {
 
-                // Sort all children nodes based on dimension value
-                List<InMemoryTreeNode> sortedChildren = new ArrayList<>(node.children.values());
-                sortedChildren.sort(
-                    Comparator.comparingInt(InMemoryTreeNode::getNodeType).thenComparingLong(InMemoryTreeNode::getDimensionValue)
-                );
-
+                int totalNumberOfChildren = 0;
                 int firstChildId = currentNodeId + queue.size() + 1;
-                int lastChildId = firstChildId + sortedChildren.size() - 1;
+
+                if (node.childStarNode != null) {
+                    totalNumberOfChildren++;
+                    queue.add(node.childStarNode);
+                }
+
+                if (node.children != null) {
+                    // Sort all children nodes based on dimension value
+                    List<InMemoryTreeNode> sortedChildren = new ArrayList<>(node.children.values());
+                    sortedChildren.sort(
```

Comment:
Can you add a TODO here based on https://github.com/opensearch-project/OpenSearch/pull/15295#discussion_r1732236197